pkg/resource/deploy/step_generator.go

```diff
@@ -1222,24 +1291,83 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, err
 	if !isTargeted {
 		sg.sames[urn] = true
 		sg.skippedCreates[urn] = true
-		return []Step{NewSkippedCreateStep(sg.deployment, event, new)}, nil
+		return []Step{NewSkippedCreateStep(sg.deployment, event, new)}, false, nil
 	}
 
 	sg.creates[urn] = true
 	logging.V(7).Infof("Planner decided to create '%v' (inputs=%v)", urn, new.Inputs)
-	return []Step{NewCreateStep(sg.deployment, event, new)}, nil
+	return []Step{NewCreateStep(sg.deployment, event, new)}, false, nil
 }
 
 func (sg *stepGenerator) generateStepsFromDiff(
 	event RegisterResourceEvent, urn resource.URN, old, new *resource.State,
 	oldInputs, oldOutputs, inputs resource.PropertyMap,
 	prov plugin.Provider, goal *resource.Goal, randomSeed []byte,
 	autonaming *plugin.AutonamingOptions,
+) ([]Step, bool, error) {
+	diff, pcs, err := sg.diff(
+		event, goal, autonaming, randomSeed,
+		urn, old, new, oldInputs, inputs, prov)
+	if pcs != nil {
+		return []Step{
+			NewDiffStep(sg.deployment, pcs, old, new, goal.IgnoreChanges),
+		}, true, nil
+	}
+
+	updateSteps, err := sg.continueStepsFromDiff(
+		event, err, diff, urn, old, new, goal, prov, autonaming, randomSeed)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if len(updateSteps) == 0 {
+		// Diff didn't produce any steps for this resource.  Fall through and indicate that it
+		// is same/unchanged.
+		logging.V(7).Infof("Planner decided not to update '%v' after diff (same) (inputs=%v)", urn, new.Inputs)
+
+		// No need to update anything, the properties didn't change.
+		sg.sames[urn] = true
+		updateSteps = []Step{NewSameStep(sg.deployment, event, old, new)}
+	}
+
+	return updateSteps, false, nil
+}
+
+func (sg *stepGenerator) ContinueStepsFromDiff(event ContinueResourceDiffEvent) ([]Step, error) {
```

Comment:
We should have a doc comment for this. Probably also `generateStepsFromDiff`, with the two explaining how they differ from each other/the use cases they meet.